pkg/ebpf/compiler/compiler.cpp

```diff
@@ -0,0 +1,273 @@
+#include "compiler.h"
+#include "files.h"
+
+#include <iostream>
+#include <llvm/IR/LegacyPassManager.h>
+#include <llvm/Support/TargetSelect.h>
+#include <clang/CodeGen/CodeGenAction.h>
+#include <clang/Driver/Job.h>
+#include <clang/Driver/Tool.h>
+#include <clang/Frontend/CompilerInstance.h>
+#include <clang/Lex/PreprocessorOptions.h>
+
+bool ClangCompiler::llvmInitialized = false;
+
+enum Architecture { PPC, PPCLE, S390X, ARM64, X86 };
+
+std::map<std::string, std::unique_ptr<llvm::MemoryBuffer>> ClangCompiler::remapped_files_;
+
+ClangCompiler::ClangCompiler(const char *name) :
+    llvmContext(new llvm::LLVMContext),
+    diagOpts(new clang::DiagnosticOptions()),
+    errStream(errString),
+    textDiagnosticPrinter(new clang::TextDiagnosticPrinter(errStream, diagOpts.get())),
+    diagnosticsEngine(new clang::DiagnosticsEngine(diagID, diagOpts, textDiagnosticPrinter.get(), false)),
+    defaultCflags({
+        "clang", // DO NOT REMOVE, first flag is ignored
+        "-emit-llvm",
+        "-O2",
+        "-D__KERNEL__",
+        "-fno-color-diagnostics",
+        "-fno-unwind-tables",
+        "-fno-asynchronous-unwind-tables",
+        "-fno-stack-protector",
+        "-nostdinc",
+        "-isystem/virtual/lib/clang/include",
+        "-x", "c"
+    }),
+    theTriple("bpf")
+{
+    if (!ClangCompiler::llvmInitialized) {
```

Comment:
Note: This code is not thread safe if 2 instances of `ClangCompiler` are created by different threads, maybe not an issue but it worths mention it.